src/common/config.c

```diff
@@ -0,0 +1,314 @@
+
+        config_save_entry_value(addr, value);
+
+        // Flash write only allows pointers in __xdata, enforce this.
+        while (len) {
+            len--;
+            flash_write_u8(addr + CONFIG_FLASH_ENTRY_NAME_OFFSET + len, config_short[len]);
+        }
+
+        return true;
+    } else {
+        return false;
+    }
+}
+
+void config_save_entries() __reentrant {
+    config_t *entry = gConfigEntries;
+
+    while (entry) {
+        (void)config_save_entry(entry->config_short, entry->value.value);
+        entry = entry->next;
+    }
+}
+
+bool config_flash_found(uint32_t addr, const char* config_short, uint8_t len) {
+    return (0 == flash_strncmp(addr + CONFIG_FLASH_ENTRY_NAME_OFFSET, config_short, len));
+}
+
+bool config_flash_valid(uint32_t addr) {
+    return (CONFIG_FLASH_ENTRY_CURRENT == flash_read_u8(addr + CONFIG_FLASH_ENTRY_CURRENT_OFFSET));
+}
+
+uint8_t config_flash_len(uint32_t addr) {
+    return flash_read_u8(addr + CONFIG_FLASH_ENTRY_LEN_OFFSET);
+}
+
+int32_t config_flash_value(uint32_t addr) {
+    return flash_read_u32(addr + CONFIG_FLASH_ENTRY_VALUE_OFFSET);
+}
+
+uint32_t condif_flash_next(uint32_t addr) {
+    uint8_t len = config_flash_len(addr);
+    return addr + CONFIG_FLASH_ENTRY_NAME_OFFSET + len;
+}
+
+uint32_t config_find_entry(const char* config_short) {
+    if(config_validate_magic()) {
+        uint32_t addr = CONFIG_ADDRESS + sizeof(config_magic);
+        uint32_t limit = addr + sizeof(config_persistant);
+
+        while(addr < limit) {
+            uint8_t len;
+            // configuration present.
+            len = config_flash_len(addr);
+
+            if(CONFIG_FLASH_ENTRY_LENGTH_INVALID == len) {
+                // no more entries exist.
+                break;
+            }
+
+            if(config_flash_valid(addr)) {
+                if(config_flash_found(addr, config_short, len)) {
+                    return addr;
+                }
+            }
+
+            // Go to the next entry.
+            addr = condif_flash_next(addr);
+        }
+    }
+
+    // Not found.
+    return 0;
+}
+
+void config_invalidate_entry(uint32_t addr) {
+    if (addr) {
+        flash_write_u8(addr + CONFIG_FLASH_ENTRY_CURRENT_OFFSET, CONFIG_FLASH_ENTRY_OBSOLETE);
+    }
+}
+
+bool config_register(config_t *entry) __reentrant {
+    // Validate entry parameters.
+    if (!entry) {
+        return false;
+    }
+
+    // Short name and desc must be set.
+    if (!entry->config_short || !entry->config_desc) {
+        return false;
+    }
+
+    // Config validated, register it
+    entry->next = gConfigEntries;
+    gConfigEntries = entry;
+
+    // Read the stored value from flash, if present.
+    uint32_t flash_addr = config_find_entry(entry->config_short);
+    if (flash_addr) {
+        int32_t saved_value = config_flash_value(flash_addr);
+
+        volatile int64_t min = entry->value.min_value;
+        volatile int64_t max = entry->value.max_value;
```

Comment:
Unnecessary volatile?

Reply:
I was seeing this improve DSEG/OSEG usage here as well. I've removed the volatile keywords the memory issue's improved in other ways since I ran into the problem.